pkg/integrations/v2/register.go

```diff
@@ -28,18 +33,52 @@ var (
 //
 // Register panics if cfg is not a pointer.
 func Register(cfg Config, ty Type) {
-	if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
-		panic(fmt.Sprintf("Register must be given a pointer, got %T", cfg))
+	// Special case: we don't need a transformer because cfg is already a Config.
+	RegisterDynamic(cfg, cfg.Name(), ty, nil)
+}
+
+// RegisterDynamic registers a dynamic integration v which does not implement
+// Config directly. The transform function will be invoked to create a wrapper
+// implementation of Config. transform will be invoked after unmarshaling v from
+// YAML, and the result will be unwrapped back to the original form at marshal
+// time.
+//
+// RegisterDynamic is only useful in niche instances. When possible, use
```

Comment:
There might be some cases where it's useful, but nothing comes to mind immediately and we can always support those later if we have to. 

I'll play around with the typing so that you explicitly have to pass an old config here rather than any value.

Reply:
As for the separate package: I did consider that, but I couldn't find a way to have normal registration be public but also have this edge case without an import loop. Changing the typing might help here

Reply:
With the name changes, and making `registerIntegration` private, I'm *MUCH* more comfortable with this. The intent is much more clear, and while the flexibility remains, it doesn't encourage consumers to use it as a catch-all.